lib/http_server.ex

```diff
@@ -15,19 +18,87 @@ defmodule HTTPServer do
   end
 
   defp serve(socket) do
+    request =
+      socket
+      |> read_line()
+      |> parse()
+
+    response =
+      request
+      |> build_response()
+      |> format_response()
+
     socket
-    |> read_line()
-    |> write_line(socket)
+    |> write_line(response)
     |> serve()
   end
 
   defp read_line(socket) do
-    {:ok, data} = :gen_tcp.recv(socket, 0)
-    data
+    case :gen_tcp.recv(socket, 0) do
+      {:ok, data} ->
+        data
+
+      {:error, :econnrefused} ->
+        exit(:normal)
+
+      {:error, :closed} ->
+        exit(:normal)
+    end
   end
 
-  defp write_line(data, socket) do
+  defp write_line(socket, data) do
     :gen_tcp.send(socket, data)
     socket
   end
+
+  def parse(message) do
+    [request_data | body] = message |> String.split("\r\n\r\n")
+    [first | headers] = request_data |> String.split("\r\n")
+    [method, path, resource] = first |> String.split(" ")
+
+    req = %Request{
+      method: method,
+      path: path,
+      resource: resource,
+      headers: parse_headers(headers, %{}),
+      body: hd(body)
+    }
+
+    case method do
+      "" ->
+        hd(body)
+
+      "POST" ->
+        %{req | method: :post}
+    end
+  end
+
+  def build_response(req) do
+    %Response{
+      status_code: 200,
+      status_message: :ok,
+      resource: req.resource,
+      headers: req.headers,
+      body: req.body
+    }
+  end
+
+  def format_response(res) do
+    "#{res.resource} #{res.status_code} OK\r\n" <>
+      "#{response_headers(res.headers)}" <>
+      "\r\n" <>
+      "#{res.body}"
+  end
+
+  def response_headers(headers) do
```

Comment:
This seems like it might be formatting headers into key-value pairs. Would `format_response_headers` be a more explicit name? Just food for thought if it makes sense in this case? 

Then it would read like: `format_response_headers(res.headers)`

Reply:
Maybe for a future PR, but I think there maybe some modules hiding out here. All the domain language is begging for it.

Reply:
the way I like to think about it is - functions --> verbs, variables --> nouns (generally speaking). so this stood out to me as a code smell because `response_headers` is a noun, but this is a function performing an action, so the name and action didn't align 😄